### PR TITLE
autotest: add the interface of ABACUS in autotest

### PR DESCRIPTION
**Pull request by pxlxingliang**

### CODE REVIEW

File: dpgen/auto_test/lib/abacus.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+             "Ru":101.07,"Rh":102.9055,"Pd":106.42,"Ag":107.8682,"Cd":112.411,"In":114.818,"Sn":118.71,"Sb":121.76,\
+             "I":126.9045,"Te":127.6,"Xe":131.293,"Cs":132.9055,"Ba":137.327,"La":138.9055,"Ce":140.116,"Pr":140.9077,\
+             "Nd":144.24,"Pm":145,"Sm":150.36,"Eu":151.964,"Gd":157.25,"Tb":158.9253,"Dy":162.5,"Ho":164.9303,"Er":167.259,\
+             "Tm":168.9342,"Yb":173.04,"Lu":174.967,"Hf":178.49,"Ta":180.9479,"W":183.84,"Re":186.207,"Os":190.23,"Ir":192.217,\
+             "Pt":195.078,"Au":196.9665,"Hg":200.59,"Tl":204.3833,"Pb":207.2,"Bi":208.9804,"Po":209,"At":210,"Rn":222,"Fr":223,\
+             "Ra":226,"Ac":227,"Pa":231.0359,"Th":232.0381,"Np":237,"U":238.0289,"Am":243,"Pu":244,"Cm":247,"Bk":247,"Cf":251,\
+             "Es":252,"Fm":257,"Md":258,"No":259,"Rf":261,"Lr":262,"Db":262,"Bh":264,"Sg":266,"Mt":268,"Rg":272,"Hs":277}
+key_words_list = ["ATOMIC_SPECIES", "NUMERICAL_ORBITAL", "LATTICE_CONSTANT", "LATTICE_VECTORS", "ATOMIC_POSITIONS", "NUMERICAL_DESCRIPTOR"]
+
+def poscar2stru(poscar,inter_param,stru):
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/lib/abacus.py:36]: Possible to call dpdata to convert the structure file?

File: dpgen/auto_test/EOS.py [hunk 2]
```diff
@@ -123,10 +126,18 @@ def make_confs(self,
...
+                if self.inter_param['type'] == 'abacus':
+                    equi_contcar = os.path.join(path_to_equi,abacus.final_stru(path_to_equi))
+                    if not os.path.isfile(equi_contcar):
+                        raise RuntimeError("Can not find %s, please do relaxation first" % equi_contcar)
+                    stru_data = abacus_scf.get_abacus_STRU(equi_contcar)  
+                    vol_to_poscar = abs(np.linalg.det(stru_data['cells'])) / np.array(stru_data['atom_numbs']).sum()
+                else:
+                    equi_contcar = os.path.join(path_to_equi, 'CONTCAR')
+                    if not os.path.exists(equi_contcar):
+                        raise RuntimeError("please do relaxation first")
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/EOS.py:139]: This lines should be reused with abacus

File: dpgen/auto_test/EOS.py [hunk 3]
```diff
@@ -138,11 +149,14 @@ def make_confs(self,
                     output_task = os.path.join(path_to_work, 'task.%06d' % task_num)
                     os.makedirs(output_task, exist_ok=True)
                     os.chdir(output_task)
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/EOS.py:151]: provide variables POSCAR = 'POSCAR' or 'STRU' and POSCAR_orig = 'POSCAR.orig' or 'STRU.orig'

File: dpgen/auto_test/Elastic.py [hunk 4]
```diff
@@ -149,26 +166,52 @@ def make_confs(self,
...
+                kp.write_file("KPOINTS")
+                os.chdir(cwd)
+                kpoints_universal = os.path.abspath(os.path.join(task_list[0], '..', 'KPOINTS'))
+                for ii in task_list:
+                    if os.path.isfile(os.path.join(ii, 'KPOINTS')):
+                        os.remove(os.path.join(ii, 'KPOINTS'))
+                    if os.path.islink(os.path.join(ii, 'KPOINTS')):
+                        os.remove(os.path.join(ii, 'KPOINTS'))
+                    os.chdir(ii)
+                    os.symlink(os.path.relpath(kpoints_universal), 'KPOINTS')
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/Elastic.py:213]: Reuse this lines with abacus.

File: dpgen/auto_test/Surface.py [hunk 5]
```diff
@@ -123,27 +126,44 @@ def make_confs(self,
...
-                # gen structure
-                ss = Structure.from_file(equi_contcar)
+
+                if self.inter_param['type'] == 'abacus':
+                    stru_data = abacus_scf.get_abacus_STRU(equi_contcar)
+                    species = []
+                    for i in range(len(stru_data['atom_numbs'])):
+                        for j in range(stru_data['atom_numbs'][i]): species.append(stru_data['atom_names'][i])
+                    ptypes = stru_data['atom_names']
+                    ss = Structure(stru_data['cells'],species,stru_data['coords'],coords_are_cartesian=True)
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/Surface.py:146]: convert abacus structure to POSCAR by dpdata and then to `Structure`.

File: dpgen/auto_test/common_equi.py [hunk 6]
```diff
@@ -103,11 +115,11 @@ def make_equi(confs,
...
                                       "relax_shape": True,
                                       "relax_vol": True}
-    elif "relax_pos" not in relax_param['cal_setting']:
+    if "relax_pos" not in relax_param['cal_setting']:
         relax_param['cal_setting']['relax_pos'] = True
-    elif "relax_shape" not in relax_param['cal_setting']:
+    if "relax_shape" not in relax_param['cal_setting']:
         relax_param['cal_setting']['relax_shape'] = True
-    elif "relax_vol" not in relax_param['cal_setting']:
+    if "relax_vol" not in relax_param['cal_setting']:
```
Comment (wanghan-iapcm) [inline dpgen/auto_test/common_equi.py:122]: @kevinwenminion plz check if the changes are fixing a bug in the code.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
@kevinwenminion where should the document of abacus+auto_test place?

**Comment by wanghan-iapcm** [comment]
Please fix the issue in the UTs

**Comment by pxlxingliang** [comment]
> Please fix the issue in the UTs

There have some difficult, I don't know how to deal with this situation. 
Some codes will call the `abacus/relax` system in dptata, which is just implemented in devel branch at last week, it seems dpgen will not  call the codes of dpdata in devel branch.

**Comment by wanghan-iapcm** [comment]
> > Please fix the issue in the UTs
> 
> There have some difficult, I don't know how to deal with this situation. Some codes will call the `abacus/relax` system in dptata, which is just implemented in devel branch at last week, it seems dpgen will not call the codes of dpdata in devel branch.

OK I will release a new version of dpdata

**Comment by wanghan-iapcm** [comment]
dpdata v0.2.8 has released. the issue in the UT is fixed. 

By PR #801 the calculation of gamma line is supported. This causes the conflict. Please also support the gamma line calculation with abacus and resolve the conflict. Thank you

**Comment by kevinwenminion** [comment]
> @kevinwenminion where should the document of abacus+auto_test place?

Maybe we can also first add the documents and examples in dpgen wiki.

File: dpgen/auto_test/common_equi.py (hunk 6 above)
Comment (kevinwenminion) [reply dpgen/auto_test/common_equi.py:122]: I don't think these three "if" can be in parallel with the previous "if" for 'cal_setting' in relax_param or not. If 'cal_setting' is not in relax_param, which means the first "if" is true, there is no meaning to go through the next three "if".

Comment (pxlxingliang) [reply dpgen/auto_test/common_equi.py:122]: @kevinwenminion 
It seems, these codes are doing initialization for the value of 'relax_pos', 'relax_shape' and 'relax_vol' in  relax_param['cal_setting']. 
We should set the value for each key.
When 'cal_setting' is in relax_param, but only set one or none of 'relax_pos', 'relax_shape' and 'relax_vol', there will leads to at least one parameter has no value.

Comment (kevinwenminion) [reply dpgen/auto_test/common_equi.py:122]: Thanks for pointing this out. Maybe we could revise the code like this:

if 'cal_setting' not in relax_param:
        relax_param['cal_setting'] = {"relax_pos": True,
                                      "relax_shape": True,
                                      "relax_vol": True}
else:
    if "relax_pos" not in relax_param['cal_setting']:
        relax_param['cal_setting']['relax_pos'] = True
    if "relax_shape" not in relax_param['cal_setting']:
        relax_param['cal_setting']['relax_shape'] = True
    if "relax_vol" not in relax_param['cal_setting']:
        relax_param['cal_setting']['relax_vol'] = True

Comment (pxlxingliang) [reply dpgen/auto_test/common_equi.py:122]: Ok, your advise is good, I will modify the code.

**Review by wanghan-iapcm — APPROVED** [review]
LGTM

**Review by kevinwenminion — APPROVED** [review]

**Comment by pxlxingliang** [comment]
> > @kevinwenminion where should the document of abacus+auto_test place?
> 
> Maybe we can also first add the documents and examples in dpgen wiki.

@kevinwenminion I have no permission to edit the wiki. Can I send you the modification, and then you do the edition?

**Comment by kevinwenminion** [comment]
> > > @kevinwenminion where should the document of abacus+auto_test place?
> > 
> > 
> > Maybe we can also first add the documents and examples in dpgen wiki.
> 
> @kevinwenminion I have no permission to edit the wiki. Can I send you the modification, and then you do the edition?
Sure, I can help to add to the wiki.

**Comment by pxlxingliang** [comment]
@kevinwenminion please check the attached file.

[Task-type.md](https://github.com/deepmodeling/dpgen/files/9314518/Task-type.md)

**Comment by kevinwenminion** [comment]
> @kevinwenminion please check the attached file.
> 
> [Task-type.md](https://github.com/deepmodeling/dpgen/files/9314518/Task-type.md)

Got it with thanks.

**Pull request merged**